compiler/luci/pass/src/QuantizeWeightsWithGPTQPass.cpp

```diff
@@ -0,0 +1,674 @@
+    for (uint32_t i = 0; i < size_hessian; i++)
+    {
+      for (uint32_t j = 0; j < i; j++)
+      {
+        float tmp = hessian[i * size_hessian + j];
+        hessian[i * size_hessian + j] = hessian[j * size_hessian + i];
+        hessian[j * size_hessian + i] = tmp;
+      }
+    }
+
+    std::vector<float> error(size);
+
+    loco::TensorShape dimension_channel_last;
+    dimension_channel_last.rank(4);
+
+    loco::TensorShape dimension_hessian;
+    dimension_hessian.rank(2);
+    dimension_hessian.dim(0).set(size_hessian);
+    dimension_hessian.dim(1).set(size_hessian);
+
+    quantize = [&](uint32_t *indices, loco::TensorShape &dimension, int channel_dim_index) {
+      quantized_values[cal_offset(dimension, indices)] = calculate_qauntized_value(
+        node, indices, dimension, channel_dim_index, scaling_factor, nudged_max, nudged_min);
+
+      uint32_t indices_channel_last[4] = {
+        indices[0], indices[3], indices[1], indices[2] // ohwi -> oihw
+      };
+
+      uint32_t dimension_channel_last[4] = {dimension.dim(0).value(), dimension.dim(3).value(),
+                                            dimension.dim(1).value(), dimension.dim(2).value()};
+
+      uint32_t idx_quant_column =
+        dimension_channel_last[2] * dimension_channel_last[3] * indices_channel_last[1] +
+        dimension_channel_last[3] * indices_channel_last[2] + indices_channel_last[3];
+
+      uint32_t indices_diag_hessian[2] = {idx_quant_column, idx_quant_column};
+
+      uint32_t channel_idx = indices[channel_dim_index];
+      auto data = node->at<loco::DataType::FLOAT32>(cal_offset(dimension, indices));
+
+      error[cal_offset(dimension, indices)] =
+        (data - (quantized_values[cal_offset(dimension, indices)] - zp[channel_idx]) *
+                  scaling_factor[channel_idx]) /
+        hessian[cal_offset_2d(dimension_hessian, indices_diag_hessian)];
+
+      if (channel_idx == (dimension.dim(channel_dim_index).value() - 1))
+      {
+        for (uint32_t o = 0; o < dimension_channel_last[0]; o++)
+        {
+          for (uint32_t i = 0; i < dimension_channel_last[1]; i++)
+          {
+            for (uint32_t h = 0; h < dimension_channel_last[2]; h++)
+            {
+              for (uint32_t w = 0; w < dimension_channel_last[3]; w++)
+              {
+                // convert coordination
+                uint32_t indices_channel_first[4] = {o, h, w, i};
+                uint32_t indices_error[4] = {o, indices[1], indices[2], indices[3]};
+                uint32_t idx_ihw = dimension_channel_last[2] * dimension_channel_last[3] * i +
+                                   dimension_channel_last[3] * h + w;
+                uint32_t indices_hessain[2] = {idx_quant_column, idx_ihw};
+
+                node->at<loco::DataType::FLOAT32>(cal_offset(dimension, indices_channel_first)) -=
+                  error[cal_offset(dimension, indices_error)] *
+                  hessian[cal_offset_2d(dimension_hessian, indices_hessain)];
```

Comment:
ditto

Reply:
I will make the changes as suggested. Thank you for your feedback 👍